specs/src/recovery.md

```diff
@@ -0,0 +1,525 @@
+essentially the same time.
+
+## Broadcast Tree
+
+In order to increase the "recovery" throughput substantially, the network must
+consider both speed, efficiency, and how much data the broadcaster must upload.
+
+The first bottleneck for throughput with a single broadcaster is the
+broadcaster's upload bandwidth.
+
+Where:
+
+- **T**: Max Theoretical Throughput (e.g., in bytes per second)
+- **B**: Upload bandwidth of the broadcaster (e.g., in bytes per second)
+- **n**: Number of times the block has to be uploaded
+
+```python
+T = B / n
+```
+
+Therefore, for recovery, to increase throughput the network should attempt to
+limit the number of times each proposer is required to upload the block data.
+
+This can be done simply by the proposer uploading only a portion to each of its
+peers. Instead of the proposer using its bandwidth to upload the same parts to
+different peers, the peers can gossip each portion to each other.
+
+```mermaid
+sequenceDiagram
+    participant NodeA as Node A
+    participant NodeB as Node B
+    participant NodeC as Node C
+
+    Note over NodeA, NodeC: time 0
+    NodeA->>NodeB: HAVE_0
+    NodeA->>NodeC: HAVE_1
+
+    Note over NodeA, NodeC: time 1
+    NodeB->>NodeA: WANT_0
+    NodeB->>NodeC: HAVE_0
+    NodeC->>NodeA: WANT_1
+    NodeC->>NodeB: HAVE_1
+
+    Note over NodeA, NodeC: time 2
+    
+    NodeA->>NodeB: DATA_0
+    NodeA->>NodeC: DATA_1
+
+    Note over NodeA, NodeC: time 3
+    NodeB->>NodeC: DATA_0
+    NodeC->>NodeB: DATA_1
+```
+
+**Figure 3: Sequence graph of the broadcast tree. Note how portions of the block
+`Data` are transferred between nodes.** Times are added to indicate which
+operation occur at essentially the same time.
+
+### Optimal Routing
+
+One of the benefits of generating the route of data on fly instead of
+pre-planning the route is that the plan can incorporate real time data. 
+
+This inherently includes:
+
+- network latency
+- as timely as possible info on the node's status
+- already distributed data
+
+However, it can also include congestion if the `Have` messages and entered into
+the send queue in a FIFO fashion as `Data`. This ensures that a haves take
+longer to be delivered to a congested connection, making it more likely peers
+will recieve that same `Have` message first from a different peer. 
+
+- **Nodes SHOULD send `Have` and `Data` messages from a FIFO queue**
+- **Nodes SHOULD request `Data` from the first peer that sends them a `Have`
+  message**
+
+Lastly, there's still a chance that a peer could already have all `Data`
+included in the `Commitment`, recreate the `Have` messages and send them to many
+nodes in the network extremely quickly. Without additional rules, this would
+cause that node to become a bottleneck in distribution, therefore we add a new
+rule to ensure this doesn't occur.
+
+- **Nodes MUST have fewer than `PerPeerConcurrentRequestCap` concurrent requests
+  to a single peer**
```

Comment:
👍  - good thinking, but wouldn't we also want to throttle/stagger it on the server side i.e. only send out x amount of `Have` messages at a time. My concern is that we'd still get congestion to that node if all it's peers (are sending `PerPeerConcurrentRequestCap` requests.